### DISCUSSION TITLE
Firestore Listener cache only updates

### DISCUSSION THREAD

**Discussion by cSieves** [invertase/react-native-firebase#8124] [Q&A]
Hi, 
I was wondering if there are any plans to allow a user to specify where to get the listener data from. At the moment as far as I understand when setting up a listener data data will be pulled from both the cache and server as changes occur. However, from what I can tell, (correct me if I am wrong) doing it this way creates a number of issues; 

- It drives up costs as every change to a query/doc results in you being charged for a read. This is bad for a few reasons, but if you manage and situation like a social media app where lots of users are updating a single doc e.g. liking a post, the cost grows exponentially. 
- Frequent updates can cause re-renders in react if you are storing the new state in a react hook. This can mess with the user experience 

I noticed in the other [SDK](https://firebase.google.com/docs/firestore/query-data/listen#:~:text=Client%20SDKs%20allow%20you%20to,reflects%20local%20data%20and%20mutations.)'s there is an option to specify where the data is coming from, is this feature going to become available in the react-native sdk soon too?

In the meantime is there some other way I can achieve the same thing i.e. get realtime updates for changes in my local cache?

**Comment by mikehardy** [comment]
Interesting, I think I see what you mean - it appears you can do a snapshot listener, but it only has the includeMetadataChanges property for listener options, it doesn't have the ability to specify cache-only

https://rnfirebase.io/reference/firestore/snapshotlistenoptions

vs

https://firebase.google.com/docs/reference/js/firestore_.snapshotlistenoptions

Tell me this isn't. confusing though - they document that property but check out their actual interface definition - https://github.com/firebase/firebase-js-sdk/blob/e3e20784503347ba5cab1ead3a7b73df87125cb4/packages/firestore-types/index.d.ts#L231-L233

Looks the same as ours!

We do allow you to specify the desired source in GetOptions for a one-time query just not the listener apparently. 

Looks like it's pretty new and that's why we don't have it (yet):
1- merged into firebase-js-sdk March 2024 https://github.com/firebase/firebase-js-sdk/pull/7982
2- merged into firebase-android-sdk about same time https://github.com/firebase/firebase-android-sdk/pull/5690
3- merged into firebase-ios-sdk about same time https://github.com/firebase/firebase-ios-sdk/pull/12370

**Reply by mikehardy** [reply]
I opened feature request #8126 just to track the request
As ever, in an open source context the quickest way to see a feature implemented is to implement the feature - otherwise patience is required, and possible understanding if its uncertain whether it will ever happen, as timing etc is out of control for anyone but the person that implements it

**Reply by cSieves** [reply]
Hi Mike, thanks for the response! 

> Tell me this isn't. confusing though - they document that property but check out their actual interface definition - https://github.com/firebase/firebase-js-sdk/blob/e3e20784503347ba5cab1ead3a7b73df87125cb4/packages/firestore-types/index.d.ts#L231-L233

Huh, yeah that is confusing 

> I opened feature request https://github.com/invertase/react-native-firebase/issues/8126 just to track the request
Unreal thank you!
